no sync between clipboard & PSE

Hi all,

I have "use platform specific extensions" & "synchronize clipboard & PSE" selected in my preferences.
When I select text, I can mouse paste it, but ctrl+v pastes something else. So I'm guessing sync isn't working?

I'm using: 
- qlipper's 5.0.0 release, https://github.com/pvanek/qlipper/archive/5.0.0.tar.gz
- lxqt 0.11 with xmonad
- linux/gentoo

I'm not seeing anything in .xsession-errors. 

Anything else you might need that could help?

Awesome, thanks! It's not building right now, but I'll track the instant_sync branch for changes.

```
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.cpp: In member function ‘void QlipperModel::indexTriggered(const QModelIndex&)’:
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.cpp:228:30: warning: extended initializer lists onlyavailable with -std=c++11 or -std=gnu++11
     list.at(row).toClipboard({QlipperItem::ToCurrent
                              ^
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.cpp:231:23: warning: extended initializer lists onlyavailable with -std=c++11 or -std=gnu++11
                     )});
                       ^
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.cpp:231:23: error: no matching function for call to ‘QlipperItem::toClipboard(<brace-enclosed initializer list>) const’
In file included from /var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.h:27:0,
                 from /var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlippermodel.cpp:25:
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlipperitem.h:62:10: note: candidate: void QlipperItem::toClipboard(const Actions&) const
     void toClipboard(const Actions & actions) const;
          ^
/var/tmp/portage/x11-misc/qlipper-9999/work/qlipper-9999/src/qlipperitem.h:62:10: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const Actions& {aka const QFlags<QlipperIte
m::Action>&}’
make[2]: *** [CMakeFiles/qlipper.dir/build.make:266: CMakeFiles/qlipper.dir/src/qlippermodel.cpp.o] Error 1
```

--- a/src/qlipperitem.cpp
+++ b/src/qlipperitem.cpp
@@ -141,8 +141,13 @@ QClipboard::Mode QlipperItem::clipBoardMode() const
     return m_mode;
 }
 
-void QlipperItem::toClipboard(bool synchronize) const
+void QlipperItem::toClipboard(const Actions & actions) const
 {
+    if (!(actions & (ToCurrent | ToOther)))
+    {
+        return;
+    }
+
     ClipboardWrap * clipboard = ClipboardWrap::Instance();
 
     QMimeData *mime = new QMimeData();
@@ -153,9 +158,14 @@ void QlipperItem::toClipboard(bool synchronize) const
         it.next();
         mime->setData(it.key(), it.value());
     }
-    clipboard->setMimeData(mime, m_mode);
-    if (synchronize && QlipperPreferences::Instance()->shouldSynchronizeClipboards())
+    if (actions.testFlag(ToCurrent))
+    {
+        clipboard->setMimeData(mime, m_mode);
+    }
+    if (actions.testFlag(ToOther))
+    {
         clipboard->setMimeData(mime, QClipboard::Clipboard == m_mode ? QClipboard::Selection : QClipboard::Clipboard);
+    }
 }
 
 QString QlipperItem::displayRole() const

--- a/src/qlipperitem.h
+++ b/src/qlipperitem.h
@@ -37,6 +37,15 @@ public:
         Sticky
     };
 
+    enum Action
+    {
+        NoAction = 0
+            , ToCurrent = 1
+            , ToOther = 1 << 1
+    };
+    Q_DECLARE_FLAGS(Actions, Action)
+
+
     QlipperItem();
     QlipperItem(QClipboard::Mode mode);
     QlipperItem(QClipboard::Mode mode, ContentType contentType, const ClipboardContent &content);
@@ -50,7 +59,7 @@ public:
     bool isValid() const { return m_valid; }
     bool enforceHistory() const { return m_enforceHistory; }
 
-    void toClipboard(bool synchronize) const;
+    void toClipboard(const Actions & actions) const;
 
     QString displayRole() const;
     QIcon decorationRole() const;

--- a/src/qlippermodel.cpp
+++ b/src/qlippermodel.cpp
@@ -142,12 +142,17 @@ void QlipperModel::clipboard_changed(QClipboard::Mode mode)
         {
             if (i->clipBoardMode() == item.clipBoardMode())
             {
-                i->toClipboard(false);
+                i->toClipboard(QlipperItem::ToCurrent);
                 m_currentIndex = index(m_sticky.count() + (i - m_dynamic.begin()));
                 break;
             }
         }
         return;
+    }
+
+    if (QlipperPreferences::Instance()->shouldSynchronizeClipboardsInstantly())
+    {
+        item.toClipboard(QlipperItem::ToOther);
     }
 
     // evaluate sticky items...
@@ -220,7 +225,9 @@ void QlipperModel::indexTriggered(const QModelIndex & index)
 
     int row = index.row();
     QList<QlipperItem> list = getList(row);
-    list.at(row).toClipboard(true);
+    QlipperItem::Actions actions(QlipperItem::ToCurrent);
+    actions |= QlipperPreferences::Instance()->shouldSynchronizeClipboards() ? QlipperItem::ToOther : QlipperItem::NoAction;
+    list.at(row).toClipboard(actions);
     m_currentIndex = index;
     if (m_sticky.size() <= index.row())
     {

--- a/src/qlipperpreferences.cpp
+++ b/src/qlipperpreferences.cpp
@@ -189,9 +189,9 @@ bool QlipperPreferences::platformExtensions() const
     return value("platformExtensions", false).toBool();
 }
 
-bool QlipperPreferences::synchronizePSE() const
+QlipperPreferences::PSESynchronization QlipperPreferences::synchronizePSE() const
 {
-    return value("synchronizePSE", false).toBool();
+    return static_cast<PSESynchronization>(value("synchronizePSE", PSE_NO_SYNC).toInt());
 }
 
 bool QlipperPreferences::clearItemsOnExit() const
@@ -222,5 +222,10 @@ int QlipperPreferences::networkPort() const
 
 bool QlipperPreferences::shouldSynchronizeClipboards() const
 {
-    return platformExtensions() && synchronizePSE();
+    return platformExtensions() && PSE_NO_SYNC != synchronizePSE();
+}
+
+bool QlipperPreferences::shouldSynchronizeClipboardsInstantly() const
+{
+    return platformExtensions() && PSE_SYNC_INSTANTLY == synchronizePSE();
 }

--- a/src/qlipperpreferences.h
+++ b/src/qlipperpreferences.h
@@ -28,6 +28,14 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 class QlipperPreferences : public QSettings
 {
 public:
+    enum PSESynchronization
+    {
+        PSE_NO_SYNC = 0
+            , PSE_SYNC_ON_SELECTION = 1
+            , PSE_SYNC_INSTANTLY = 2
+    };
+
+public:
     static QlipperPreferences *Instance();
     ~QlipperPreferences();
 
@@ -41,7 +49,7 @@ public:
     QString shortcut() const;
     int historyCount() const;
     bool platformExtensions() const;
-    bool synchronizePSE() const;
+    PSESynchronization synchronizePSE() const;
     bool clearItemsOnExit() const;
     bool synchronizeHistory() const;
 
@@ -50,6 +58,7 @@ public:
     int networkPort() const;
 
     bool shouldSynchronizeClipboards() const;
+    bool shouldSynchronizeClipboardsInstantly() const;
 
 private:
     QlipperPreferences();

--- a/src/qlipperpreferencesdialog.cpp
+++ b/src/qlipperpreferencesdialog.cpp
@@ -37,7 +37,7 @@ QlipperPreferencesDialog::QlipperPreferencesDialog(QWidget *parent) :
     const bool pse = s->platformExtensions();
     platformExtensionsCheckBox->setChecked(pse);
     synchronizePSE->setEnabled(pse);
-    synchronizePSE->setChecked(s->synchronizePSE());
+    synchronizePSE->setCurrentIndex(s->synchronizePSE());
     clearItemsOnExit->setChecked(s->clearItemsOnExit());
     synchronizeHistory->setChecked(s->synchronizeHistory());
 
@@ -83,7 +83,7 @@ void QlipperPreferencesDialog::accept()
     s->setValue("displaySize", displaySizeComboBox->value());
     s->setValue("trim", trimCheckBox->isChecked());
     s->setValue("platformExtensions", platformExtensionsCheckBox->isChecked());
-    s->setValue("synchronizePSE", synchronizePSE->isChecked());
+    s->setValue("synchronizePSE", synchronizePSE->currentIndex());
     s->setValue("shortcut", shortcutWidget->keySequence().toString());
     s->setValue("clearItemsOnExit", clearItemsOnExit->isChecked());
     s->setValue("synchronizeHistory", synchronizeHistory->isChecked());

--- a/src/qlipperpreferencesdialog.ui
+++ b/src/qlipperpreferencesdialog.ui
@@ -148,13 +148,6 @@
          </property>
         </widget>
        </item>
-       <item row="6" column="0" colspan="2">
-        <widget class="QCheckBox" name="synchronizePSE">
-         <property name="text">
-          <string>Synchronize clipboard &amp;&amp; PSE when item selected</string>
-         </property>
-        </widget>
-       </item>
        <item row="7" column="0" colspan="2">
         <widget class="QCheckBox" name="clearItemsOnExit">
          <property name="text">
@@ -167,6 +160,25 @@
          <property name="text">
           <string>Synchronize history to storage instantly</string>
          </property>
+        </widget>
+       </item>
+       <item row="6" column="0" colspan="2">
+        <widget class="QComboBox" name="synchronizePSE">
+         <item>
+          <property name="text">
+           <string>No synchronization of clipboard &amp; PSE</string>
+          </property>
+         </item>
+         <item>
+          <property name="text">
+           <string>Synchronize clipboard &amp; PSE when item selected</string>
+          </property>
+         </item>
+         <item>
+          <property name="text">
+           <string>Synchronize clipboard &amp; PSE instantly</string>
+          </property>
+         </item>
         </widget>
        </item>
       </layout>

